File: C-Plus-Plus/ds/Queue_using_stack.cpp
/*
A queue can be implemented using two stacks. 
Let queue to be implemented be q and stacks used to implement q be s1 and s2.
*/

#include <iostream>
#include <stack>
using namespace std;


class Queue{

public:

	//main data structure
	stack<int> s1;				
	//helper
	stack<int> s2;				

	void push(int x){ 
		s1.push(x);
	}

	void pop(){

		//copy s1 to s2

		while(!s1.empty()){
			s2.push(s1.top());
			s1.pop();
		}

		s2.pop();

		//copy s2 to s1

		while(!s2.empty()){
			s1.push(s2.top());
			s2.pop();
		}

	}

	int front(){
		//copy s1 to s2

		while(!s1.empty()){
			s2.push(s1.top());
			s1.pop();
		}

		int x = s2.top();

		//copy s2 to s1

		while(!s2.empty()){
			s1.push(s2.top());
			s2.pop();
		}

		return x;

	}

	int size(){
		return s1.size();
	}

	bool empty(){
		return s1.empty();
	}


};




int main()
{
	Queue q;

	int n;
	cin >> n;

	for(int i=1;i<=n;i++){
		q.push(i);
	}


	cout<<q.front()<<endl;

	cout<<q.size()<<endl;

	q.pop();
	cout<<q.front()<<endl;
	
	cout<<q.empty()<<endl;


	while(!q.empty()){
		cout<<q.front()<<endl;
		q.pop();
	}


	
	return 0;
}

/* 
	Test case :
	Input : 5
	Output : 
	1
	5
	2
	0
	2
	3
	4
	5
	Time Complexity : 
	Push operation : O(N).
	Pop operation : O(1).
	SpaceComplexity : O(N).
*/

File: Java/cp/Queue_using_Stack.java

// Java program to implement Queue using two stacks with costly enQueue()  
import java.util.*; 
  
class QueuetoStack  
{  
static class Queue  
{  
    static Stack<Integer> stk1 = new Stack<Integer>();  
    static Stack<Integer> stk2 = new Stack<Integer>();  
  
    static void enQueue(int x)  
    {  
        // Move all elements from stk1 to stk2  
        while (!stk1.isEmpty()) 
        {  
            stk2.push(stk1.pop());  
            //stk1.pop();  
        }  
  
        // Push item into stk1  
        stk1.push(x);  
  
        // Push everything back to stk1  
        while (!stk2.isEmpty())  
        {  
            stk1.push(stk2.pop());  
            //stk2.pop();  
        }  
    }  
  
    // Dequeue an item from the queue  
    static int deQueue()  
    {  
        // if first stack is empty  
        if (stk1.isEmpty())  
        {  
            System.out.println("Q is Empty");  
            System.exit(0);  
        }  
  
        // Return top of stk1  
        int s = stk1.peek();  
        stk1.pop();  
        return s;  
    }  
};  
  
// Main method 
public static void main(String[] args)  
{  
    Scanner sc = new Scanner(System.in);
    System.out.println("Enter limit");
    int n = sc.nextInt();
    Queue qu = new Queue();
    for(int i = 1; i <= n; i++)
       qu.enQueue(i);  
    for(int i = 1; i <= n; i++)
       System.out.println(qu.deQueue());  
    
}  
}


/* Time Complexity:
  1) Push operation: O(N).
  2) Pop operation: O(1).

Space Complexity: O(N).

Input:
10

Output:
1
2
3
4
5
6
7
8
9
10
*/
